src/sinks/util/tcp.rs

```diff
@@ -37,82 +39,182 @@ use tokio_util::codec::{BytesCodec, FramedWrite};
+
+    pub fn build_service(&self, cx: SinkContext) -> crate::Result<(TcpService, Healthcheck)> {
+        let connector = self.build_connector(cx)?;
+        let healthcheck = connector.healthcheck();
+        Ok((connector.into(), healthcheck))
+    }
+
+    pub fn build(
+        &self,
+        cx: SinkContext,
+        encoding: EncodingConfig<Encoding>,
+    ) -> crate::Result<(VectorSink, Healthcheck)> {
+        let connector = self.build_connector(cx.clone())?;
+        let healthcheck = connector.healthcheck();
+        let sink: TcpSink = connector.into();
+        let sink = StreamSinkOld::new(sink, cx.acker())
+            .with_flat_map(move |event| iter_ok(encode_event(event, &encoding)));
+
+        Ok((VectorSink::Futures01Sink(Box::new(sink)), healthcheck))
+    }
+}
+
+impl TcpConnector {
+    fn new(host: String, port: u16, resolver: Resolver, tls: MaybeTlsSettings) -> Self {
+        Self {
+            host,
+            port,
+            resolver,
+            tls,
+        }
+    }
+
+    fn connect(&self) -> BoxFuture<'static, Result<TcpOrTlsStream, TcpError>> {
+        let host = self.host.clone();
+        let port = self.port;
+        let resolver = self.resolver;
+        let tls = self.tls.clone();
+
+        async move {
+            let ip = resolver
+                .lookup_ip(host.clone())
+                .await
+                .context(DnsError)?
+                .next()
+                .ok_or(TcpError::NoAddresses)?;
+
+            let addr = SocketAddr::new(ip, port);
+            let stream = tls.connect(host, addr).await.context(ConnectError)?;
+            Ok(FramedWrite::new(stream, BytesCodec::new()))
+        }
+        .boxed()
+    }
+
+    fn healthcheck(&self) -> BoxFuture<'static, crate::Result<()>> {
+        self.connect().map_ok(|_| ()).map_err(|e| e.into()).boxed()
+    }
+}
+
+impl Into<TcpSink> for TcpConnector {
+    fn into(self) -> TcpSink {
+        TcpSink::new(self.host, self.port, self.resolver, self.tls)
+    }
+}
+
+impl Into<TcpService> for TcpConnector {
+    fn into(self) -> TcpService {
+        TcpService { connector: self }
+    }
+}
+
+pub struct TcpService {
+    connector: TcpConnector,
+}
+
+impl tower::Service<Bytes> for TcpService {
+    type Response = ();
+    type Error = TcpError;
+    type Future = BoxFuture<'static, Result<(), Self::Error>>;
 
-        let encoding = self.encoding.clone();
-        let sink = Box::new(
-            StreamSinkOld::new(tcp, cx.acker())
-                .with_flat_map(move |event| iter_ok(encode_event(event, &encoding))),
-        );
+    fn poll_ready(&mut self, _: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        Poll::Ready(Ok(()))
+    }
 
-        Ok((VectorSink::Futures01Sink(sink), healthcheck))
+    fn call(&mut self, msg: Bytes) -> Self::Future {
+        use futures::SinkExt;
+        let connector = self.connector.clone();
+        async move {
+            let mut connection = connector.connect().await?;
```

Comment:
Is this ok that connection created for each call?

Reply:
Yeah this is the reason for moving to StreamSink https://github.com/timberio/vector/pull/3557#issuecomment-703045855

Reply:
okay, but is this required by statsd? Currently, udp socket bind once and then only send data.

Reply:
Statsd does not say anything about how many tcp connections are opened, so I don't think it cares.

Reply:
We should not be re-resolving DNS and opening a new connection for every single event that we're forwarding. That is going to perform very poorly.